File: frontend/src/services/pedidoService.js

const API_URL = 'http://localhost:8080/api/pedidos';

// Obtener usuarioId del localStorage
const getUsuarioId = () => {
    let usuarioId = localStorage.getItem('usuarioId');
    
    if (!usuarioId) {
        usuarioId = localStorage.getItem('userId');
    }
    
    if (!usuarioId) {
        const usuario = localStorage.getItem('usuario');
        if (usuario) {
            try {
                const usuarioObj = JSON.parse(usuario);
                usuarioId = usuarioObj.id;
            } catch (e) {
                console.error('Error al parsear usuario:', e);
            }
        }
    }
    
    console.log('Usuario ID encontrado:', usuarioId);
    
    if (!usuarioId) {
        throw new Error('Debe iniciar sesión');
    }
    
    // Convertir a string para asegurar consistencia
    return String(usuarioId);
};

// Obtener token JWT del localStorage
const getToken = () => {
    return localStorage.getItem('token');
};

// Crear headers con autenticación
const getHeaders = () => {
    const headers = {
        'Content-Type': 'application/json'
    };
    
    const token = getToken();
    if (token) {
        headers['Authorization'] = `Bearer ${token}`;
    }
    
    try {
        const usuarioId = getUsuarioId();
        if (usuarioId) {
            headers['Usuario-Id'] = usuarioId;
        }
    } catch (e) {
        // No agregar Usuario-Id si no está logueado
    }
    
    console.log('Headers generados:', headers);
    
    return headers;
};

// Verificar si el usuario está logueado
export const estaLogueado = () => {
    try {
        const usuarioId = getUsuarioId();
        const token = getToken();
        return !!(usuarioId && token);
    } catch (error) {
        return false;
    }
};

// Obtener carrito del usuario
export const obtenerCarrito = async () => {
    try {
        const headers = getHeaders();
        console.log('Obteniendo carrito con headers:', headers);
        
        const response = await fetch(`${API_URL}/carrito`, {
            headers: headers
        });

        console.log('Respuesta obtenerCarrito:', response.status);

        if (response.status === 401) {
            throw new Error('Debe iniciar sesión');
        }

        if (!response.ok) {
            const errorText = await response.text();
            console.error('Error del servidor:', errorText);
            throw new Error(errorText || 'Error al obtener el carrito');
        }

        return await response.json();
    } catch (error) {
        console.error('Error en obtenerCarrito:', error);
        throw error;
    }
};

// Agregar licencia al carrito
export const agregarLicencia = async (licenciaId) => {
    try {
        const usuarioId = getUsuarioId();
        console.log('Agregando licencia', licenciaId, 'para usuario:', usuarioId);
        
        const headers = getHeaders();
        console.log('Headers para agregar licencia:', headers);
        
        const response = await fetch(`${API_URL}/carrito/licencia/${licenciaId}`, {
            method: 'POST',
            headers: headers
        });

        console.log('Respuesta agregarLicencia:', response.status);

        if (response.status === 401) {
            throw new Error('Debe iniciar sesión');
        }

        if (!response.ok) {
            const error = await response.text();
            console.error('Error del servidor:', error);
            throw new Error(error || 'Error al agregar al carrito');
        }

        const result = await response.json();
        console.log('Licencia agregada exitosamente:', result);
        return result;
    } catch (error) {
        console.error('Error en agregarLicencia:', error);
        throw error;
    }
};

// Eliminar licencia del carrito
export const eliminarLicencia = async (licenciaId) => {
    try {
        const response = await fetch(`${API_URL}/carrito/licencia/${licenciaId}`, {
            method: 'DELETE',
            headers: getHeaders()
        });

        if (!response.ok) {
            throw new Error('Error al eliminar del carrito');
        }

        return await response.json();
    } catch (error) {
        console.error('Error en eliminarLicencia:', error);
        throw error;
    }
};

// Vaciar carrito
export const vaciarCarrito = async () => {
    try {
        const response = await fetch(`${API_URL}/carrito/vaciar`, {
            method: 'DELETE',
            headers: getHeaders()
        });

        if (!response.ok) {
            throw new Error('Error al vaciar el carrito');
        }

        return true;
    } catch (error) {
        console.error('Error en vaciarCarrito:', error);
        throw error;
    }
};

// Confirmar compra
export const confirmarCompra = async (cantidades) => {
    try {
        console.log('Enviando cantidades al backend:', cantidades);
        
        const response = await fetch(`${API_URL}/carrito/confirmar`, {
            method: 'POST',
            headers: getHeaders(),
            body: JSON.stringify(cantidades)
        });

        console.log('Respuesta confirmarCompra:', response.status);

        if (response.status === 401) {
            throw new Error('Debe iniciar sesión');
        }

        if (!response.ok) {
            const errorData = await response.json().catch(() => ({}));
            throw new Error(errorData.message || 'Error al confirmar la compra');
        }

        const result = await response.json();
        console.log('Compra confirmada:', result);
        return result;
    } catch (error) {
        console.error('Error al confirmar compra:', error);
        throw error;
    }
};

// Obtener historial de pedidos
export const obtenerHistorial = async () => {
    try {
        const response = await fetch(`${API_URL}/historial`, {
            headers: getHeaders()
        });

        if (!response.ok) {
            throw new Error('Error al obtener el historial');
        }

        return await response.json();
    } catch (error) {
        console.error('Error en obtenerHistorial:', error);
        throw error;
    }
};

// Actualizar cantidad de una licencia en el carrito
export const actualizarCantidad = async (licenciaId, cantidad) => {
    try {
        console.log(`Actualizando cantidad de licencia ${licenciaId} a ${cantidad}`);
        
        const response = await fetch(`${API_URL}/carrito/licencia/${licenciaId}/cantidad?cantidad=${cantidad}`, {
            method: 'PUT',
            headers: getHeaders()
        });

        console.log('Respuesta actualizarCantidad:', response.status);

        if (response.status === 401) {
            throw new Error('Debe iniciar sesión');
        }

        if (!response.ok) {
            const error = await response.text();
            throw new Error(error || 'Error al actualizar cantidad');
        }

        const result = await response.json();
        console.log('Cantidad actualizada:', result);
        return result;
    } catch (error) {
        console.error('Error en actualizarCantidad:', error);
        throw error;
    }
};

File: backend/src/main/java/com/crud/backend/service/PedidoService.java
package com.crud.backend.service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.crud.backend.model.Licencia;
import com.crud.backend.model.Pedido;
import com.crud.backend.model.PedidoLicencia;
import com.crud.backend.model.Usuario;
import com.crud.backend.repository.LicenciaRepository;
import com.crud.backend.repository.PedidoRepository;
import com.crud.backend.repository.UsuarioRepository;

@Service
public class PedidoService {

    @Autowired
    private PedidoRepository pedidoRepository;

    @Autowired
    private LicenciaRepository licenciaRepository;

    @Autowired
    private UsuarioRepository usuarioRepository;

    // Obtener o crear el carrito (pedido único) del usuario
    @Transactional
    public Pedido obtenerCarrito(Long usuarioId) {
        System.out.println("Buscando carrito para usuario ID: " + usuarioId);

        Optional<Pedido> carritoExistente = pedidoRepository
            .findByUsuarioIdAndEstado(usuarioId, "carrito");

        if (carritoExistente.isPresent()) {
            System.out.println("Carrito existente encontrado con " + carritoExistente.get().getPedidoLicencias().size() + " licencias");
            return carritoExistente.get();
        }

        System.out.println("Creando nuevo carrito para usuario ID: " + usuarioId);

        Usuario usuario = usuarioRepository.findById(usuarioId)
            .orElseThrow(() -> {
                System.err.println("Usuario no encontrado con ID: " + usuarioId);
                return new RuntimeException("Usuario no encontrado");
            });

        Pedido nuevoCarrito = new Pedido();
        nuevoCarrito.setUsuario(usuario);
        nuevoCarrito.setEstado("carrito");
        nuevoCarrito.setTotal(0.0);

        Pedido carritoGuardado = pedidoRepository.save(nuevoCarrito);
        System.out.println("Nuevo carrito creado con ID: " + carritoGuardado.getId());

        return carritoGuardado;
    }

    // Agregar licencia al carrito
    @Transactional
    public Pedido agregarLicencia(Long usuarioId, Long licenciaId) {
        System.out.println("Agregando licencia ID " + licenciaId + " al carrito del usuario ID " + usuarioId);

        Pedido carrito = obtenerCarrito(usuarioId);

        Licencia licencia = licenciaRepository.findById(licenciaId)
            .orElseThrow(() -> {
                System.err.println("Licencia no encontrada con ID: " + licenciaId);
                return new RuntimeException("Licencia no encontrada");
            });

        // Verificar stock
        if (licencia.getCantidad_stock() <= 0) {
            System.err.println("Sin stock para licencia: " + licencia.getNombre());
            throw new RuntimeException("No hay stock disponible para esta licencia");
        }

        // Verificar si la licencia ya está en el carrito
        boolean yaExiste = carrito.getPedidoLicencias().stream()
            .anyMatch(pl -> pl.getLicencia().getId().equals(licenciaId));

        if (yaExiste) {
            System.out.println("La licencia ya está en el carrito");
            calcularTotal(carrito);
            Pedido carritoActualizado = pedidoRepository.save(carrito);
            return carritoActualizado;
        }

        PedidoLicencia pedidoLicencia = new PedidoLicencia();
        pedidoLicencia.setPedido(carrito);
        pedidoLicencia.setLicencia(licencia);
        pedidoLicencia.setCantidad(1);

        carrito.getPedidoLicencias().add(pedidoLicencia);
        calcularTotal(carrito);

        Pedido carritoActualizado = pedidoRepository.save(carrito);
        System.out.println("Licencia agregada. Total de licencias: " + carritoActualizado.getPedidoLicencias().size());
        System.out.println("Total del carrito: $" + carritoActualizado.getTotal());

        return carritoActualizado;
    }

    // Eliminar licencia del carrito
    @Transactional
    public Pedido eliminarLicencia(Long usuarioId, Long licenciaId) {
        System.out.println("Eliminando licencia ID " + licenciaId + " del carrito del usuario ID " + usuarioId);

        Pedido carrito = obtenerCarrito(usuarioId);

        boolean eliminada = carrito.getPedidoLicencias().removeIf(pl -> pl.getLicencia().getId().equals(licenciaId));

        if (eliminada) {
            System.out.println("Licencia eliminada del carrito");
        } else {
            System.out.println("Licencia no estaba en el carrito");
        }

        calcularTotal(carrito);

        Pedido carritoActualizado = pedidoRepository.save(carrito);
        System.out.println("Total de licencias restantes: " + carritoActualizado.getPedidoLicencias().size());
        System.out.println("Total del carrito: $" + carritoActualizado.getTotal());

        return carritoActualizado;
    }

    // Vaciar carrito
    @Transactional
    public void vaciarCarrito(Long usuarioId) {
        System.out.println("Vaciando carrito del usuario ID: " + usuarioId);

        Pedido carrito = obtenerCarrito(usuarioId);
        int cantidadLicencias = carrito.getPedidoLicencias().size();

        carrito.getPedidoLicencias().clear();
        carrito.setTotal(0.0);
        pedidoRepository.save(carrito);

        System.out.println("Carrito vaciado. Se eliminaron " + cantidadLicencias + " licencias");
    }

    // Confirmar compra y descontar stock
    @Transactional
    public Pedido confirmarCompra(Long usuarioId) {
        System.out.println("Confirmando compra del usuario ID: " + usuarioId);

        Pedido carrito = obtenerCarrito(usuarioId);

        if (carrito.getPedidoLicencias().isEmpty()) {
            System.err.println("El carrito está vacío");
            throw new RuntimeException("El carrito está vacío");
        }

        System.out.println("Verificando stock de " + carrito.getPedidoLicencias().size() + " licencias...");

        // Verificar stock de todas las licencias
        for (PedidoLicencia pl : carrito.getPedidoLicencias()) {
            Licencia licencia = pl.getLicencia();
            int cantidadSolicitada = pl.getCantidad();
            if (licencia.getCantidad_stock() < cantidadSolicitada) {
                System.err.println("Stock insuficiente para: " + licencia.getNombre());
                throw new RuntimeException("No hay stock disponible para: " + licencia.getNombre());
            }
        }

        // DESCONTAR STOCK
        System.out.println("Descontando stock:");
        for (PedidoLicencia pl : carrito.getPedidoLicencias()) {
            Licencia licencia = pl.getLicencia();
            int cantidadSolicitada = pl.getCantidad();

            int nuevoStock = licencia.getCantidad_stock() - cantidadSolicitada;
            licencia.setCantidad_stock(nuevoStock);
            licenciaRepository.save(licencia);

            System.out.println("- " + licencia.getNombre() + ": " + licencia.getCantidad_stock() + " → " + nuevoStock);
        }

        // Cambiar estado del carrito a confirmado
        carrito.setEstado("confirmado");
        Pedido pedidoConfirmado = pedidoRepository.save(carrito);

        System.out.println("Compra confirmada. Pedido ID: " + pedidoConfirmado.getId());
        System.out.println("Total de la compra: $" + pedidoConfirmado.getTotal());

        return pedidoConfirmado;
    }

    // Obtener historial de pedidos confirmados del usuario
    public List<Pedido> obtenerHistorialPedidos(Long usuarioId) {
        System.out.println("Obteniendo historial de pedidos del usuario ID: " + usuarioId);

        List<Pedido> historial = pedidoRepository.findByUsuarioId(usuarioId).stream()
            .filter(p -> !"carrito".equals(p.getEstado()))
            .collect(Collectors.toList());

        System.out.println("Se encontraron " + historial.size() + " pedidos confirmados");

        return historial;
    }

    public Optional<Pedido> getById(Long pedidoId) {
        return pedidoRepository.findById(pedidoId);
    }

    @Transactional
    public Pedido save(Pedido pedido) {
        System.out.println("Guardando pedido...");
        System.out.println("Pedido recibido: " + pedido);

        // Validar que las licencias no sean null
        if (pedido.getPedidoLicencias() == null || pedido.getPedidoLicencias().isEmpty()) {
            System.err.println("El pedido debe tener al menos una licencia");
            throw new IllegalArgumentException("El pedido debe tener al menos una licencia");
        }

        System.out.println("Licencias en el pedido: " + pedido.getPedidoLicencias().size());

        // Calcular total automáticamente
        calcularTotal(pedido);

        // Estado por defecto
        if (pedido.getEstado() == null || pedido.getEstado().isEmpty()) {
            pedido.setEstado("PENDIENTE");
        }

        Pedido guardado = pedidoRepository.save(pedido);
        System.out.println("Pedido guardado con ID: " + guardado.getId());
        System.out.println("Total: $" + guardado.getTotal());

        return guardado;
    }

    @Transactional
    public void delete(Long id) {
        System.out.println("Eliminando pedido ID: " + id);
        pedidoRepository.deleteById(id);
        System.out.println("Pedido eliminado");
    }

    // Calcular total
    private void calcularTotal(Pedido pedido) {
        double total = pedido.getPedidoLicencias().stream()
            .mapToDouble(pl -> pl.getLicencia().getPrecio() * pl.getCantidad())
            .sum();
        pedido.setTotal(total);
    }

    @Transactional
    public Pedido actualizarCantidadLicencia(Long usuarioId, Long licenciaId, Integer cantidad) {
        System.out.println("Actualizando cantidad para usuario " + usuarioId +
                        ", licencia " + licenciaId + ", cantidad " + cantidad);

        Pedido carrito = obtenerCarrito(usuarioId);

        if (cantidad <= 0) {
            throw new RuntimeException("La cantidad debe ser mayor a 0");
        }

        // Buscar el PedidoLicencia correspondiente
        PedidoLicencia pedidoLicencia = carrito.getPedidoLicencias().stream()
            .filter(pl -> pl.getLicencia().getId().equals(licenciaId))
            .findFirst()
            .orElseThrow(() -> new RuntimeException("La licencia no está en el carrito"));

        pedidoLicencia.setCantidad(cantidad);

        // Recalcular total
        calcularTotal(carrito);

        Pedido carritoActualizado = pedidoRepository.save(carrito);

        System.out.println("Cantidad actualizada. Nuevo total: " + carritoActualizado.getTotal());

        return carritoActualizado;
    }
}